Error while checking build/O2/fullCI for 290c55424e06a65e964628ed69a7f31f99173180 at 2023-10-24 12:17:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/12134-slc8_x86-64/0/Common/MathUtils/src/MathUtilsLinkDef.h:44:2: error: invalid preprocessing directive
Error: /sw/slc8_x86-64/ROOT/v6-28-04-alice2-5/bin/rootcling: Linkdef compilation failure
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12134/290c55424e06a65e964628ed69a7f31f99173180/build_O2_fullCI/pretty.html).

--- a/Common/MathUtils/CMakeLists.txt
+++ b/Common/MathUtils/CMakeLists.txt
@@ -16,6 +16,7 @@ o2_add_library(
           src/Chebyshev3D.cxx
           src/Chebyshev3DCalc.cxx
    src/SymMatrixSolver.cxx
+   src/Tsallis.cxx
   PUBLIC_LINK_LIBRARIES
     ROOT::Hist
     FairLogger::FairLogger
@@ -37,7 +38,8 @@ o2_target_root_dictionary(
           include/MathUtils/RandomRing.h
           include/MathUtils/Primitive2D.h
           include/MathUtils/SMatrixGPU.h
-          include/MathUtils/SymMatrixSolver.h)
+          include/MathUtils/SymMatrixSolver.h
+          include/MathUtils/Tsallis.h)
 
 o2_add_test(
   CachingTF1

--- a/Common/MathUtils/include/MathUtils/Tsallis.h
+++ b/Common/MathUtils/include/MathUtils/Tsallis.h
@@ -1,0 +1,48 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef ALICEO2_TSALLIS_H
+#define ALICEO2_TSALLIS_H
+
+namespace o2
+{
+
+namespace math_utils
+{
+
+struct Tsallis {
+  /// Tsallis/Hagedorn function describing charged pt spectra (m s = 62.4 GeV to 13 TeV) as in https://iopscience.iop.org/article/10.1088/2399-6528/aab00f/pdf
+  /// https://github.com/alisw/AliPhysics/blob/523f2dc8b45d913e9b7fda9b27e746819cbe5b09/PWGPP/AliAnalysisTaskFilteredTree.h#L145
+  /// \param pt     - transverse momentum
+  /// \param mass   - mass of particle
+  /// \param sqrts  - centre of mass energy
+  /// \return       - invariant yields of the charged particle *pt
+  ///    n(sqrts)= a + b/sqrt(s)                             - formula 6
+  ///    T(sqrts)= c + d/sqrt(s)                             - formula 7
+  ///    a = 6.81 ± 0.06       and b = 59.24 ± 3.53 GeV      - for charged particles page 3
+  ///    c = 0.082 ± 0.002 GeV and d = 0.151 ± 0.048 (GeV)   - for charged particles page 4
+  static float tsallisCharged(float pt, float mass, float sqrts);
+
+  /// Random downsampling trigger function using Tsallis/Hagedorn spectra fit (sqrt(s) = 62.4 GeV to 13 TeV) as in https://iopscience.iop.org/article/10.1088/2399-6528/aab00f/pdf
+  /// \return flat q/pt trigger
+  /// \param pt pat of particle
+  /// \param factorPt defines the sampling
+  /// \param sqrts centre of mass energy
+  /// \param weight weight which is internally calculated
+  /// \param rnd random value between (0->1) used to check for sampling
+  /// \param mass particles mass (use pion if not known)
+  static bool downsampleTsallisCharged(float pt, float factorPt, float sqrts, float& weight, float rnd, float mass = 0.13957);
+};
+
+} // namespace math_utils
+} // namespace o2
+
+#endif

--- a/Common/MathUtils/src/MathUtilsLinkDef.h
+++ b/Common/MathUtils/src/MathUtilsLinkDef.h
@@ -41,5 +41,6 @@
 #pragma link C++ class o2::math_utils::IntervalXYd_t + ;
 #pragma link C++ class o2::math_utils::Bracketf_t + ;
 #pragma link C++ class o2::math_utils::Bracketd_t + ;
+#pragma link C++ class o2::math_utils::Tsallis + ;
 
 #endif

--- a/Common/MathUtils/src/Tsallis.cxx
+++ b/Common/MathUtils/src/Tsallis.cxx
@@ -1,0 +1,39 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "MathUtils/Tsallis.h"
+#include <cmath>
+
+namespace o2::math_utils
+{
+
+float Tsallis::tsallisCharged(float pt, float mass, float sqrts)
+{
+  const float a = 6.81;
+  const float b = 59.24;
+  const float c = 0.082;
+  const float d = 0.151;
+  const float mt = std::sqrt(mass * mass + pt * pt);
+  const float n = a + b / sqrts;
+  const float T = c + d / sqrts;
+  const float p0 = n * T;
+  return std::pow((1. + mt / p0), -n) * pt;
+}
+
+bool Tsallis::downsampleTsallisCharged(float pt, float factorPt, float sqrts, float& weight, float rnd, float mass)
+{
+  const float prob = tsallisCharged(pt, mass, sqrts);
+  const float probNorm = tsallisCharged(1., mass, sqrts);
+  weight = prob / probNorm;
+  return (rnd * (weight * pt * pt)) < factorPt;
+}
+
+} // namespace o2::math_utils

--- a/Common/Utils/include/CommonUtils/DebugStreamer.h
+++ b/Common/Utils/include/CommonUtils/DebugStreamer.h
@@ -50,7 +50,7 @@ enum SamplingTypes {
   sampleID = 2,       ///< sample every n IDs (per example track)
   sampleIDGlobal = 3, ///< in case different streamers have access to the same IDs use this gloabl ID
   sampleWeights = 4,  ///< perform sampling on weights, defined where the streamer is called
-  sampleTsalis = 5,   ///< perform sampling on tsalis pdf
+  sampleTsallis = 5,  ///< perform sampling on tsallis pdf
 };
 
 #if !defined(GPUCA_GPUCODE) && !defined(GPUCA_STANDALONE)
@@ -164,34 +164,12 @@ class DebugStreamer
   /// \return returns integer index for given streamer flag
   static int getIndex(const StreamFlags streamFlag);
 
-  /// Random downsampling trigger function using Tsalis/Hagedorn spectra fit (sqrt(s) = 62.4 GeV to 13 TeV) as in https://iopscience.iop.org/article/10.1088/2399-6528/aab00f/pdf
-  /// \return flat q/pt trigger
-  /// \param pt pat of particle
-  /// \param factorPt defines the sampling
-  /// \param sqrts centre of mass energy
-  /// \param weight weight which is internally calculated
-  /// \param rnd random value between (0->1) used to check for sampling
-  /// \param mass particles mass (use pion if not known)
-  static bool downsampleTsalisCharged(float pt, float factorPt, float sqrts, float& weight, float rnd, float mass = 0.13957);
-
   /// get random value between min and max
   static float getRandom(float min = 0, float max = 1);
 
  private:
   using StreamersPerFlag = tbb::concurrent_unordered_map<size_t, std::unique_ptr<o2::utils::TreeStreamRedirector>>;
   StreamersPerFlag mTreeStreamer; ///< streamer which is used for the debugging
-
-  /// Tsalis/Hagedorn function describing charged pt spectra (m s = 62.4 GeV to 13 TeV) as in https://iopscience.iop.org/article/10.1088/2399-6528/aab00f/pdf
-  /// https://github.com/alisw/AliPhysics/blob/523f2dc8b45d913e9b7fda9b27e746819cbe5b09/PWGPP/AliAnalysisTaskFilteredTree.h#L145
-  /// \param pt     - transverse momentum
-  /// \param mass   - mass of particle
-  /// \param sqrts  - centre of mass energy
-  /// \return       - invariant yields of the charged particle *pt
-  ///    n(sqrts)= a + b/sqrt(s)                             - formula 6
-  ///    T(sqrts)= c + d/sqrt(s)                             - formula 7
-  ///    a = 6.81 ± 0.06       and b = 59.24 ± 3.53 GeV      - for charged particles page 3
-  ///    c = 0.082 ± 0.002 GeV and d = 0.151 ± 0.048 (GeV)   - for charged particles page 4
-  static float tsalisCharged(float pt, float mass, float sqrts);
 #else
 
   // empty implementation of the class for GPU or when the debug streamer is not build for CPU

--- a/Common/Utils/src/DebugStreamer.cxx
+++ b/Common/Utils/src/DebugStreamer.cxx
@@ -107,29 +107,6 @@ bool o2::utils::DebugStreamer::checkStream(const StreamFlags streamFlag, const s
   return true;
 }
 
-float o2::utils::DebugStreamer::tsalisCharged(float pt, float mass, float sqrts)
-{
-  const float a = 6.81;
-  const float b = 59.24;
-  const float c = 0.082;
-  const float d = 0.151;
-  const float mt = std::sqrt(mass * mass + pt * pt);
-  const float n = a + b / sqrts;
-  const float T = c + d / sqrts;
-  const float p0 = n * T;
-  const float result = std::pow((1. + mt / p0), -n) * pt;
-  return result;
-}
-
-bool o2::utils::DebugStreamer::downsampleTsalisCharged(float pt, float factorPt, float sqrts, float& weight, float rnd, float mass)
-{
-  const float prob = tsalisCharged(pt, mass, sqrts);
-  const float probNorm = tsalisCharged(1., mass, sqrts);
-  weight = prob / probNorm;
-  const bool isSampled = (rnd * (weight * pt * pt)) < factorPt;
-  return isSampled;
-}
-
 float o2::utils::DebugStreamer::getRandom(float min, float max)
 {
   // init random number generator for each thread

--- a/Detectors/TPC/workflow/src/TPCTimeSeriesSpec.cxx
+++ b/Detectors/TPC/workflow/src/TPCTimeSeriesSpec.cxx
@@ -29,6 +29,7 @@
 #include "TPCCalibration/RobustAverage.h"
 #include "DetectorsCalibration/IntegratedClusterCalibrator.h"
 #include "CommonUtils/DebugStreamer.h"
+#include "MathUtils/Tsallis.h"
 #include "ReconstructionDataFormats/TrackTPCITS.h"
 #include "CommonDataFormat/AbstractRefAccessor.h"
 #include "ReconstructionDataFormats/PrimaryVertex.h"
@@ -780,8 +781,8 @@ class TPCTimeSeries : public Task
         const float factorPt = sampling.second;
         bool writeData = true;
         float weight = 0;
-        if (sampling.first == o2::utils::SamplingTypes::sampleTsalis) {
-          writeData = o2::utils::DebugStreamer::downsampleTsalisCharged(tracksTPC[iTrk].getPt(), factorPt, mSqrt, weight, o2::utils::DebugStreamer::getRandom());
+        if (sampling.first == o2::utils::SamplingTypes::sampleTsallis) {
+          writeData = o2::math_utils::Tsallis::downsampleTsallisCharged(tracksTPC[iTrk].getPt(), factorPt, mSqrt, weight, o2::utils::DebugStreamer::getRandom());
         }
         if (writeData) {
           auto clusterMask = makeClusterBitMask(track);

